unit_tests/test_aggregator.py

```diff
@@ -0,0 +1,415 @@
+    """Includes methods which test get_uuid_from_area_name method."""
+
+    @staticmethod
+    def test_get_uuid_from_area_name__not_empty_dict(aggregator_explicit):
+        aggregator_explicit.area_name_uuid_mapping = {"TestArea": ["test_area_uuid"]}
+        area_name = "TestArea"
+
+        patch("gsy_e_sdk.aggregator.get_uuid_from_area_name_in_tree_dict",
+              return_value="test_area_uuid")
+        assert aggregator_explicit.get_uuid_from_area_name(area_name) == "test_area_uuid"
+
+    @staticmethod
+    def test_get_uuid_from_area_name__empty_dict(aggregator_explicit):
+        aggregator_explicit.area_name_uuid_mapping = {}
+        area_name = "TestArea"
+
+        patch("gsy_e_sdk.aggregator.get_uuid_from_area_name_in_tree_dict",
+              return_value="test_area_uuid")
+        assert aggregator_explicit.get_uuid_from_area_name(area_name) is None
+
+
+class TestCalculateGridFee:
+    """Includes methods which test calculate_grid_fee method."""
+
+    @staticmethod
+    def test_calculate_grid_fee__called(aggregator_explicit):
+        with patch("gsy_e_sdk.aggregator.GridFeeCalculation.calculate_grid_fee") as mocked_method:
+            aggregator_explicit.calculate_grid_fee("start_market",
+                                                   "target_market",
+                                                   "current_market_fee")
+            mocked_method.assert_called_with("start_market",
+                                             "target_market",
+                                             "current_market_fee")
+
+    @staticmethod
+    def test_calculate_grid_fee__return(aggregator_explicit):
+        with patch("gsy_e_sdk.aggregator.GridFeeCalculation.calculate_grid_fee",
+                   return_value="test_ok"):
+            ret_val = aggregator_explicit.calculate_grid_fee("start_market",
+                                                             "target_market",
+                                                             "current_market_fee")
+            assert ret_val == "test_ok"
+
+
+class TestExecuteBatchCommands:
+    """Includes methods which test execute_batch_commands method."""
+
+    @staticmethod
+    def test_execute_batch_commands__command_buffer_length_zero(aggregator_explicit):
+        with patch("gsy_e_sdk.aggregator.ClientCommandBuffer.buffer_length",
+                   new_callable=PropertyMock, return_value=0):
+            assert aggregator_explicit.execute_batch_commands() is None
+
+    @staticmethod
+    def test_execute_batch_commands__execute_batch_called(aggregator_explicit):
+        with patch("gsy_e_sdk.aggregator.ClientCommandBuffer.execute_batch") as mocked_method:
+            aggregator_explicit.start_websocket_connection()
+            aggregator_explicit.execute_batch_commands()
+            mocked_method.assert_called_once()
+
+    @staticmethod
+    def test_execute_batch_commands__validate_all_uuids_in_list(aggregator_explicit):
```

Comment:
A similar thing as [this](https://github.com/gridsingularity/gsy-e-sdk/pull/194/files#r764782518) applies here. `_all_uuids_in_selected_device_uuid_list` is a private method which is not seen from the "outside", so our tests should not know anything about its existence. This would belong to the category of "private query", which can just be ignored when testing. 

Testing this method binds us to the implementation of the class (while we should test its interface). If this method doesn't work properly, we will have other tests failing because of it and that's already enough (it avoids tests repetition).

Reply:
This is well explained in the talk I mentioned, specifically in the section about query/command "sent to self" (private).

What are we testing here? What we care about is that, when a device UUID is not in `self.device_uuid_list` (which is part of the public interface), calling `execute_batch_commands()` will raise an `Exception`. `_all_uuids_in_selected_device_uuid_list()` is just an implementation detail, it's the name of a method that does part of what I described above. But if someday I change the name of that method my tests shouldn't fail. Also, if I just move its content inside `execute_batch_commands()`, my tests shouldn't fail.
Tests should only fail if the exception is not raised when the device is not in the list, or if the exception is raised when it is.

Let me know if this explains a bit or if you have other concerns about it. These topics are interesting and complex in my opinion, so it's nice to discuss them if necessary.

Reply:
I find your comment interesting because it pictures something I mentioned in a previous comment.

The main test case is devoted to test execute_batch_commands() logic. In order to completely do this there are multiple test uses for example this one we're commenting on which asserts whether this private method was called or not. I understand that it would be implementation binding to make this assertion. 

However, I understand you are suggesting a test in which `_all_uuids_in_selected_device_uuid_list()` logic gets tested, but (correctly if I'm wrong) test it as a part of `execute_batch_commands()` logic. I mean that, IMO whether an exception is raised or not depends on `_all_uuids_in_selected_device_uuid_list()` working properly and not on `execute_batch_commands()` logic. 
I can see this behavior is somehow "inherited" from `_all_uuids_in_selected_device_uuid_list()` to `execute_batch_commands()`, but shouldn't this be tested on say `test_all_uuids_in_selected_device_uuid_list()`.

In other words, with your suggestion, wouldn't we be implicitly testing `_all_uuids_in_selected_device_uuid_list()` behavior when we should be exclusively testing `execute_batch_commands()` logic?

This relates to [this](https://github.com/gridsingularity/gsy-e-sdk/pull/194#discussion_r767962175) comment. In your opinion, how should that work?

Reply:
> In other words, with your suggestion, wouldn't we be implicitly testing _all_uuids_in_selected_device_uuid_list() behavior when we should be exclusively testing execute_batch_commands() logic?

Yes, we would be indirectly testing `_all_uuids_in_selected_device_uuid_list()` and that's fine, because it is a private ("sent to self") query. We don't care about this method, we only care about our interface methods. If they work, then everything is fine for us. And, if they work, it must mean that also the private methods work, so testing them would be redundant and would bind us to the "hidden" implementation, making our tests slower and less flexible.
Private methods can be tested in some scenarios (e.g. when they're complex). But most of the time, if a private method is very complex, it probably means that it should actually be the public method of a different class.

Another way to see this: you could easily delete `_all_uuids_in_selected_device_uuid_list()` and paste its content inside `execute_batch_commands()`. This wouldn't change anything in the way `execute_batch_commands()` works, so our tests should not fail because of this. But they will fail, instead, if you make assertions or mock the call to `_all_uuids_in_selected_device_uuid_list()` and this doesn't make sense for us, because our software works just as well without it.